Add cbmc-viewer

markrtuttle opened the pull request:
This is a first pass at adding cbmc-viewer source and regression tests.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add cbmc-viewer

This is a first pass at adding cbmc-viewer source and regression tests.

Commit:
Convert cbmc-viewer to Python 3

tautschnig commented on .travis.yml:
```diff
@@ -3,6 +3,21 @@ language: cpp
+        apt:
+          packages:
+            - python
+	    - python-pip
```
@markrtuttle Github suggests that there is a tabs vs spaces mixup here.

yumibagge commented:
I am quite intrigued about this new script. I'd like to know what it does and how to use it... but is it out of scope for this PR to include README.md?

Commit:
Convert regression test.py to Python3 and remove pip install future

Commit:
Regression Makefile invokes test.py with python3

Commit:
Make cbmc-viewer regressions pass clang formatter test.

Commit:
Make cbmc-viewer print more diagnostic information.

Commit:
Add error handling to cbmc-viewer regression  subprocess.Popen calls

chrisr-diffblue commented:
Could you update the PR description to include a brief introduction to what cbmc-viewer is? That could also form the start of a README file as well.

chrisr-diffblue commented on .travis.yml:
```diff
@@ -3,6 +3,23 @@ language: cpp
+        pip3 install lxml
+      script:
+        which goto-cc
+	which cbmc
```
I'm not sure this will work - this change to the travis config adds a whole new job which doesn't do anything to actually build cbmc or its tools, so these tools won't be on the path. Depending on how long you expect it to take to run these new regression tests, it might be a better idea to integrate them into one of the other existing jobs as part of the standard regression tests?

markrtuttle replied:
I am still figuring out how to write the regression tests.  I don't know how to debug this other than on an open PR.  I created a draft PR, but travis wouldn't run on a draft PR.

markrtuttle replied:
We have done as you advised, and travis is now working.

hannes-steffenhagen-diffblue commented on regression/cbmc-viewer/test.py:
```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python3
+"""
+Regession testing of CBMC-viewer.
```
This could use some explanation of how this works and what exactly it is testing; also this seems to use a completely different testing approach to all our other regression tests, you should definitely include what motivates this difference

markrtuttle replied:
We have added some text: The script is guarding against a simple textual diff of the html that would fail if cbmc were to make some reasonable change that simply changed the text of the trace produced.  The script is parsing the html and comparing only the shapes of the result.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/Makefile:
```diff
@@ -0,0 +1,29 @@
+default:
+	@Nothing to do
```
why?

markrtuttle replied:
This Makefile has been cleaned up and this target is gone.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/Makefile:
```diff
@@ -0,0 +1,29 @@
+	$(RM) *.pyc
+	$(RM) *~ .*~ \#* .\#*
+
+veryclean: clean
```
what?

NlightNFotis replied:
There might have been a design consideration for a cleaning target that only removes bytecode cache files, and another that does some deeper cleaning, but it never materialised.

I agree that this appears to be useless, hence should be better if it is removed.

markrtuttle replied:
This Makefile has been cleaned up and this target is gone.

markrtuttle commented:
This is still a work in progress and not yet ready for review.  I've taken it out of draft mode only to get travis to run.

chrisr-diffblue commented:
Ah! That's fine - I'll add some labels to indicate that - feel free to remove the labels whenever you think it's ready for wider review.

hannes-steffenhagen-diffblue reviewed (commented):
I also second that this needs additional documentation for what this is, why it is useful and how to use it.

karkhaz committed:
Clean up cbmc-viewer regression test suite

This commit makes multiple fixes to the test script:

- Use safer subprocess invocations (without shelling out)
- Use logging for output instead of print, and log both to a file and
  stderr
- Allow tests to be located in nested subdirectories. All tests under
  the 'pass'/'fail' directory are expected to pass/fail.
- Add a CMake build script
- Add the ability to run standalone tests to the top-level CMake
  manifest (i.e. tests that don't need to be run through test.pl)

karkhaz committed:
actually add cbmc-viewer test directory

Commit:
Add readme file to cbmc-viewer

Commit:
Fix clang formatting errors in cbmc-viewer regression tests.

chrisr-diffblue commented on scripts/cbmc-viewer/errors.py:
```diff
@@ -0,0 +1,139 @@
+
+    # pylint: disable=too-many-branches
+
+    html = ['<div style="font-family:courier">']
```
Inserting this comment at a fairly random location, but it's a general comment regarding the HTML output - rather than specifying style info in hard coded ways such as this local `font-family` tag - could you do it all on a style sheet (even if the stylesheet is embedded in the generated HTML) - that would allow other tools to potentially adjust the style to their own local conventions in a much more flexible way - possibly even allowing various JavaScript DOM hackery to take place by other clients (I'm not a web programmer, so excuse my ignorance of exactly what is doable there).

markrtuttle replied:
This is added to the TODO.md list of things to do.

NlightNFotis commented on regression/cbmc-viewer/test.py:
```diff
@@ -0,0 +1,150 @@
+Regession testing of CBMC-viewer.
+"""
+
+from __future__ import print_function     # get print function from Python 3
```
Why is this importing `print_function` from `__future__` when the shebang at the top indicates that the default python version expected to run this is python 3?

markrtuttle replied:
You are reviewing an old commit of a version intended to work with both Python2 and Python3.  The second commit in this pull request makes the cut over to straight Python3 and deleted these reference to future, etc.

NlightNFotis commented on scripts/cbmc-viewer/cbmc-quiet-xml:
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/python -u
```
There's a general mixing of the shebangs here. Some are indicating that `python3` is expected to run the scripts and some are indicating that `python2` will, but with forward portability.

Is this by design? Why not default to python 3? If there's expectation that python 2 needs to be supported, why are some files asking the environment modifier to default to python 3?

markrtuttle replied:
These are old scripts used with an old project, and they have been deleted.

NlightNFotis commented on scripts/cbmc-viewer/tags.py:
```diff
@@ -0,0 +1,166 @@
+import re
+import sys
+import errno
+from past.builtins import cmp
```
This is imported but not used?

markrtuttle replied:
Again, this was removed by the conversion to straight Python3 in the second commit of this PR.

NlightNFotis commented on scripts/cbmc-viewer/tags.py:
```diff
@@ -0,0 +1,166 @@
+        self.define = {}
+        self.function = {}
+        self.variable = {}
+        self.make_tags_file(srcdir, tagscmd, tagsfile, fltr)
```
This function seems capable to raise exceptions that are not handled anywhere, therefore crashing the program. Is that by design?

markrtuttle replied:
We have added a generic exception handler to cbmc-viewer. We have added improved exception handling to the TODO.md list of things to fix in the future.

Commit:
Add python3 to linux regression, fix syntax in regression/CMakeLists.txt.

NlightNFotis commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,275 @@
+                                                                srcdir)
+
+        if not xml:
+            return
```
Also, my understanding is that quite a few functions here have failure modes that are not being handled properly, and they just fail silently. Is this what should happen?

markrtuttle replied:
We have added a generic exception handler to cbmc-viewer.  We have added improved exception handling to the TODO.md list of things to fix in the future.

NlightNFotis reviewed (commented):
Some notes thus far.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fbc33a6980b220d6273b01d1681a9d1705b67eb1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124053710
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/cbmc-quiet-xml:
```diff
@@ -0,0 +1,65 @@
+    for regexp in REGEXP:
+        if re.match(regexp,string):
+            return True
+    return False
```
this file seems to have a bunch of copy&pasted code from the other cbmc-quiet

markrtuttle replied:
This is an old script has has been removed.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/cbmc-quiet:
```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/python -u
+
+from __future__ import print_function
```
what's this file for

markrtuttle replied:
This script has been deleted.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/cbmc-viewer:
```diff
@@ -0,0 +1,6 @@
+
+import report
+
+if __name__ == "__main__":
```
why isn't this just in report?

markrtuttle replied:
Removed this wrapper script and renamed report.py -> cbmc-viewer.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/errors.py:
```diff
@@ -0,0 +1,139 @@
+    """Map property violations in result to source line of the violation."""
+    report = {}
+    report[PROP] = {}
+    report[PROP][ERROR] = []
```
why are these maps/lists if the set of things you can put into here is so limited?

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/errors.py:
```diff
@@ -0,0 +1,139 @@
+
+    # pylint: disable=too-many-branches
+
+    html = ['<div style="font-family:courier">']
```
Generally not a good way to generate html by hand like this, consider using a templating engine or at least a HTML builder to clear things up

markrtuttle replied:
Added to the TODO.md list of things to do in the near future.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/linestatus.py:
```diff
@@ -0,0 +1,31 @@
+def combine(status1, status2):
+    """Combine line status.  A HIT and a MISS combine to PARTIAL."""
+    if status1 is None and status2 is None:
+        return None
```
this is redundant

markrtuttle replied:
Done.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/linestatus.py:
```diff
@@ -0,0 +1,31 @@
+        return status1
+
+    if PARTIAL in [status1, status2]:
+        return PARTIAL
```
this is also redundant

markrtuttle replied:
Done.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/storm.py:
```diff
@@ -0,0 +1,191 @@
+"""The results of cbmc-storm."""
```
What's cbmc-storm?

markrtuttle replied:
Deleted all references to cbmc-storm tool that has not been sufficiently maintained.

"CBMC Storm, a python based tool that automates common CBMC usage. CBMC is a model checker for C programs. In cases when running CBMC in a large codebase such as an operating system is infeasible  CBMC Storm can be used to automate  CBMC runs on specific entry points of interest."

Commit:
Cleanup cbmc-viewer Makefile, remove old scripts, add TODO.md

Commit:
Remove CBMC Storm from cbmc-viewer

Commit:
Add quote=False to invocations of html.escape to correct code markup

Commit:
Add a generic exception handler to cbmc-viewer.

We wrap the main function of cbmc-viewer with a try/except that prints
a stack trace along with instructions to the user on how to submit an
issue to github if the trace does not help them debug the problem
themselves.

This commit also adds a bit of document to the top of the python script
that implements the cbmc-viewer regression testing.

Commit:
Remove cbmc-viewer wrapper script: Rename report.py as cbmc-viewer

Commit:
Simplify linestatus combination, add 'use html generator' to TODO.md list.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 6d22c26f6d670ed06738194bcfc9759364c2663a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124117481
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0cb22c6887e1ae3465c427b751fe433029a80491).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/124118778
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

hannes-steffenhagen-diffblue commented on regression/CMakeLists.txt:
```diff
@@ -20,6 +20,20 @@ macro(add_test_pl_tests cmdline)
+# Standalone regression tests that do not need test.pl
+macro(add_standalone_test cmdline)
+  add_test(
+    NAME "${name}-CORE"
```
:no_entry_sign: 
This macro doesn't have a `name` argument so this doesn't actually work. I'd recommend checking out [`cmake_parse_arguments`](https://cmake.org/cmake/help/v3.10/command/cmake_parse_arguments.html) to improve this, or at least add a `name` argument to the macro; I imagine you'd want something like

```
macro(add_standalone_test name)
add_test(
  NAME "${name}-CORE"
  COMMAND ${ARGN}  #lack of quotes intentional, this is a list argument
  WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
endmacro(add_standalone_test)
```

but tbh it'd probably be easier to just skip the macro and use `add_test` and `set_tests_properties` directly for this case; They have reasonable enough interfaces and it'll be easier to follow than custom macros.

hannes-steffenhagen-diffblue commented on regression/cbmc-viewer/.gitignore:
```diff
@@ -0,0 +1,6 @@
+TAGS
```
why a separate .gitignore for this directory?

chrisr-diffblue replied:
@markrtuttle I'd agree, this seems like it can just go in the root level .gitignore

hannes-steffenhagen-diffblue commented on regression/cbmc-viewer/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+if(NOT WIN32)
+  add_standalone_test(python3 "${CMAKE_CURRENT_SOURCE_DIR}/test.py")
```
I imagine the intention was that this should be a test that invokes `python3` with the `test.py` args. As of now this just calls `python3` with no args and ignores the second argument, see above for a suggestion for a fix.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+                                                                srcdir)
+
+        if not xml:
+            return
```
The concern with these cases is that these seem to be failures, but their location and reasons get masked by the return. Might be better for now to just raise an exception instead?

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+        for goal in root.iter("goal"):
+            goal_description = goal.get("description")
+            if not goal_description:
+                break
```
Is this an error?

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+            loc_func = goal_loc.get("function")
+            loc_line = int(goal_loc.get("line"))
+
+            lnst = (linestatus.HIT
```
nitpick: Common abbreviations like `src` and arguably `loc`/`func` are fine, but I'd avoid introducing new or uncommon ones for words like `linestatus`.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+                locations = [(loc_file, loc_func, loc_line)]
+
+            for loc in locations:
+                src, func, line = loc
```
This `src` here actually seems to refer to a filepath, the name should reflect that. Same with most (all?) instances of `src` in this file, consider renaming them to `source_file` or something along those lines.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+                    line_status[src][func] = {}
+                if not line_status[src][func].get(line):
+                    line_status[src][func][line] = None
+                line_status[src][func][line] \
```
While somewhat esoteric (although not at all unheard of), function definitions can be split over multiple files. This however seems to assume a hierarchy where functions are subordinate to files?

danielsn replied:
I'm curious about this. When would this ever be a good idea?

hannes-steffenhagen-diffblue replied:
> I'm curious about this. When would this ever be a good idea?

I've personally used this to make writing an OpenGL wrapper easier; In CBMC we use a similar technique (albeit not for function definitions, but for enums etc) to pregenerate parts of our irep_ids table.

Again, this isn't super common obviously, but it can happen in some cases; Neither the C nor the C++ standard forbids it at any rate even if you don't find it particularly useful.

hannes-steffenhagen-diffblue commented on scripts/cbmc-viewer/block.py:
```diff
@@ -0,0 +1,274 @@
+
+                if not line_status.get(src):
+                    line_status[src] = {}
+                if not line_status[src].get(func):
```
There are a lot of cases like this where it's "look this thing up in a map, if not initialise it to this, and then regardless of whether or not it was there before do this..." 

these should just be replaced with something like

```
line_status_file = line_status.setdefault(file, {})
line_status_file_function = line_status_file.setdefault(func, {})
```

etc, or otherwise be replace with utility classes with proper accessors that do that sort of bookkeeping.

hannes-steffenhagen-diffblue reviewed (commented):
Small comments. Mostly nitpicks / things which I think don't necessarily need to happen in this PR, except for the cmake thing

xbauch commented on scripts/cbmc-viewer/srcloc.py:
```diff
@@ -0,0 +1,160 @@
+"""Methods for maniuplating source locations found in cbmc output."""
```
manipulating

xbauch commented on scripts/cbmc-viewer/srcloc.py:
```diff
@@ -0,0 +1,160 @@
+def parse_xml_string(xml, abort=True, src=""):
+    """Open and parse an xml string."""
+
+    # Messages printed by cbmc before the coverage data we care about may
```
This comment seems redundant as this function does not replace `\0`.

xbauch commented on scripts/cbmc-viewer/trace.py:
```diff
@@ -0,0 +1,129 @@
+        if name:
+            self.trace[name] = trace
+
+        self.markup = markup
```
Aren't we instantiating this twice?

Commit:
Add concurrency, configuration, and cbmc warning reporting.

Add a warning section to the report for warnings issued by cbmc.
Report on functions with missing function bodies, and report
separately on other warnings (like ignoring byte-enable...).

Enable concurrent invocation of cbmc-viewer.  Use the process id to
generate a unique name for the TAGS file so concurrent invocations of
the script will not try to use the same TAGS file.  The right solution
is to use the tempfile module to get unique files in a temporary
directory, but this will not work because etags writes paths to the
TAGS file that are relative from the TAGS file to the source file.

Add a --config flag to specify a json file (default cbmc-viewer.py)
giving viewer configuration information.  Ultimately, all command line
arguments could be specified here.  For now, the only entry is
'expected-missing-functions' which is a list of names of function we
expect to have a missing function body.  If this file can be read, the
list of missing functions above is split into "expected" and
"unexpected" missing functions.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 78180bdda8dda51459a34934ab2c5f577007a14f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/135974340
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 068522a634d04b744d23d514fb3b49a12d4d41af).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136042029
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Patch cbmc-viewer reachable module for source locations without line numbers.

Commit:
Correct cbmc-viewer patch for source locations without line numbers.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8fba65ccc966eaab27b21be1af4c9324bccccafc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/139376504
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Repair cmake macro add_standalone_test definition and invocation.

This is a cmake macro defined to facilitate the cbmc-viewer regression
testing.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7fb206c819550653ad55baf042aa3585610dd15d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/139411792
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Add Windows batch script to run cbmc-viewer on Windows

Commit:
Add sources module to cbmc-viewer.

Add a sources module that uses find or walk to construct the list
of source files under the source root.  Use walk on Windows and
find everywhere else (find does not exist on Windows).  Modify tags
and markup modules to use the new sources module.

Commit:
Generate path regular expressions for cbmc-viewer on Windows.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 28e0ea47ab0a5d8a7106fc310f6a3a03251ff155).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/139442950
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

chrisr-diffblue commented on .travis.yml:
```diff
@@ -89,7 +89,7 @@ jobs:
     - stage: Linter + Doxygen + non-debug Ubuntu/gcc-5 test
       os: linux
-      dist: trusty
+      dist: xenial
```
Why do you need to bump the distro version?

hannes-steffenhagen-diffblue commented on regression/CMakeLists.txt:
```diff
@@ -20,6 +20,20 @@ macro(add_test_pl_tests cmdline)
+macro(add_standalone_test name cmdline)
+  add_test(
+    NAME "${name}-CORE"
+    COMMAND "${cmdline}"
```
Note this still doesn’t quite work like it should – this is just the first parameter passed to the macro, which is just going to be `python3`. You’ll want to remove the `cmdline` parameter completely and just use `${ARGN}` (i.e. the remainder of the arguments after the named ones) here, otherwise the other arguments (i.e. `test.py` in your case) won’t be passed on the command line.

chrisr-diffblue commented:
@peterschrammel @kroening  - Are the changes to the travis config in this PR, in particular bumping the Ubuntu version, OK? I'm personally OK with them, but just wanted to check if you knew of any users we need to support on the older Ubuntu platform?

chrisr-diffblue commented:
@markrtuttle - I may be mistaken, but as far as I can see the cbmc-viewer regression tests don't actually get run automatically? Looks like there are two reasons for this:
* For Makefile based builds, you'll need to add the cbmc-viewer directory to the `DIRS` list in `cbmc/regression/Makefile`
* For CMake based builds, you'll need to address Hannes comment https://github.com/diffblue/cbmc/pull/4863#discussion_r354405280

Commit:
Make cbmc-viewer run etags iteratively.

The command to run etags on the list of source files can generate
a command longer than shells allow (eg, on Windows).  Instead, run
etags iteratively on the source files on chunks of 100 files at a time.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9ac82ee3f620452bce061502a1fbfbc3b387cd00).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/139597837
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Use Python 3.5 subprocess API in cbmc-viewer tags module

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 61c503b58da574afcbade987ef3abb8c218ee466).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/139622862
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

karkhaz committed:
Fix add_standalone_test function for cbmc-viewer

This commit fixes the add_standalone_test CMake function. It now uses
cmake_parse_arguments.

karkhaz committed:
move cbmc-viewer .gitignore to top level

karkhaz committed:
Enable cbmc-viewer tests to run using Make

Commit:
Repair cbmc-viewer regression test pass/pass-1.

Prior set of "correct" results failed to count 3 lines of code as
reachable code.

xbauch commented:
@karkhaz I'm not a python expert but it seems that the `test.py` script for cbmc-viewer regression tests is calling `subprocess.run` with parameter `text`. Support for `text` was added in python 3.7. Which version is the CI running on?

chrisr-diffblue commented:
@xbauch @karkhaz - At least on Travis, it looks like its picking up python3.5:
```
make[1]: Leaving directory '/home/travis/build/diffblue/cbmc/regression'
22273:	Traceback (most recent call last):
22283:	  File "test.py", line 284, in <module>
22293:	    main()
22303:	  File "test.py", line 265, in main
22313:	    generateReport(options)
22323:	  File "test.py", line 196, in generateReport
22333:	    stderr=subprocess.PIPE, text=True)
22343:	  File "/usr/lib/python3.5/subprocess.py", line 693, in run
22353:	    with Popen(*popenargs, **kwargs) as process:
22363:	TypeError: __init__() got an unexpected keyword argument 'text'
22373:	make[2]: *** [test] Error 1
22383:	make[1]: *** [cbmc-viewer] Error 1
```

chrisr-diffblue commented:
And all the codebuild failures seem to be this error:
```
Running cbmc-viewer... 
make -C "cbmc-viewer" test || exit 1; 
make[2]: Entering directory '/codebuild/output/src436711421/src/github.com/diffblue/cbmc/regression/cbmc-viewer' 
Traceback (most recent call last): 
  File "test.py", line 21, in <module> 
    from   lxml    import etree               # parse HTML files 
ModuleNotFoundError: No module named 'lxml' 
Makefile:5: recipe for target 'test' failed 
```

Which I think you can fix by adding a `pip3 install lxml` to the various buildspec*.xml files.

hannes-steffenhagen-diffblue commented on regression/cbmc-viewer/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+if(NOT WIN32)
```
This seems to be redundant, as this is already checked in the parent CMakeLists. Remove one of the conditionals please.

Commit:
Repair cbmc-viewer regression failures in code build.

Added Python3 and lxml module to the code build jobs in addition to
travis.  Backed up cbmc-viewer/test.py to Python 3.5.  Cleaned up
cbmc-viewer/CMakeLists.txt.

hannes-steffenhagen-diffblue commented:
Travis is failing because of a reference to `goto-cc` - we don’t put binaries on `$PATH` during regression testing, so you’ll have to reference them via relative path in the regression Makefile

Commit:
Fix two Windows bugs in cbmc-viewer.

First, use os.path.join to build absolute paths to source files.

Second, consider as reachable functions only those functions in files
under the source root.  Reports on Windows may otherwise include as
reachable some inlined functions in system header files that are
unused by the code under test.

Commit:
Allow cbmc-viewer --srcdir to be a relative path.

Prior to this change, an error was possible if srcdir and blddir were
not given as absolute paths.  This change parses the command line
arguments and calls os.path.abspath on srcdir and blddir.

Commit:
Make cbmc-viewer top-level exception handler pass on SystemExit.

Commit:
Add cbmc tool paths to the environment of cbmc-viewer regression tests.

Commit:
Use os.pathsep to update PATH in cbmc-viewer regression

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c7565460d8e97ddc276eb05944abe043dde4d4fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144633198
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 762ea969fb3e87406d1d305bcf3b2dfe00ee4e1a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144638445
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9bca497e0f563562e76040849c13b2b4248d8841).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144642586
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

markrtuttle commented:
/bot run checks

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 47da4d254bb786a417c81055b2bfebe474368ff4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144656705
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

chrisr-diffblue commented:
@markrtuttle @tautschnig It looks like the three failing AWS codebuild linux builds (cbmc, cbmc-codecov and cbmc-linux-clang) are all failing with the following output:

```
Running cbmc-viewer...
make -C "cbmc-viewer" test \|\| exit 1;
make[2]: Entering directory '/codebuild/output/src423296307/src/github.com/diffblue/cbmc/regression/cbmc-viewer'
Unable to compare generated report with expected report
files in cwd '/codebuild/output/src423296307/src/github.com/diffblue/cbmc/regression/cbmc-viewer/pass/pass-1': [
./proofs.goto
./property.xml
./source.c
./source.goto
./cbmc-args
./harness.c
./coverage.xml
./harness.h
./cbmc.log
./compare/loop.html
./compare/harness.c.html
./compare/index.html
./compare/source.c.html
./compare/harness.h.html
]
Makefile:5: recipe for target 'test' failed
make[2]: *** [test] Error 1
make[2]: Leaving directory '/codebuild/output/src423296307/src/github.com/diffblue/cbmc/regression/cbmc-viewer'
Makefile:63: recipe for target 'cbmc-viewer' failed
make[1]: *** [cbmc-viewer] Error 1
make[1]: Leaving directory '/codebuild/output/src423296307/src/github.com/diffblue/cbmc/regression'
Makefile:56: recipe for target 'test' failed
make: *** [test] Error 1
make: Leaving directory '/codebuild/output/src423296307/src/github.com/diffblue/cbmc/regression'
```

And it looks like the cbmc-windows build failed it's test because it couldn't find python3:
```
make -C "cbmc-viewer" test \|\| exit 1;
make[2]: Entering directory '/cygdrive/c/codebuild/tmp/output/src383070643/src/github.com/diffblue/cbmc/regression/cbmc-viewer'
make[2]: python3: Command not found
make[2]: *** [Makefile:5: test] Error 127
make[2]: Leaving directory '/cygdrive/c/codebuild/tmp/output/src383070643/src/github.com/diffblue/cbmc/regression/cbmc-viewer'
make[1]: *** [Makefile:64: cbmc-viewer] Error 1
make[1]: Leaving directory '/cygdrive/c/codebuild/tmp/output/src383070643/src/github.com/diffblue/cbmc/regression'
make: *** [Makefile:56: test] Error 1
make: Leaving directory '/cygdrive/c/codebuild/tmp/output/src383070643/src/github.com/diffblue/cbmc/regression'
```

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 47da4d254bb786a417c81055b2bfebe474368ff4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144656705
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Update buildspec-windows.yml to install python3 for cbmc-viewer regression.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 2e0810362c85147277835c185468e8157e531a89).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/144838053
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

hannes-steffenhagen-diffblue commented:
```
make -C "cbmc-viewer" test \|\| exit 1;
982 | make[2]: Entering directory '/cygdrive/c/codebuild/tmp/output/src782953303/src/github.com/diffblue/cbmc/regression/cbmc-viewer'
983 | Traceback (most recent call last):
984 | File "test.py", line 21, in <module>
985 | from   lxml    import etree               # parse HTML files
986 | ModuleNotFoundError: No module named 'lxml'
```

Looks like it’s just missing a `pip install lxml` somewhere.

markrtuttle commented:
@hannes-steffenhagen-diffblue These code build issues are being debugged in my own AWS accounts, I won't be submitting an update until I've pushed through all the issues.

markrtuttle commented:
The issue with lxml is that there were multiple pythons installed.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: be88c82850952e057b56e03ba8c013415357a5cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/145848041
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

Commit:
Fix cbmc-viewer configuration file extension.

markrtuttle commented:
Closing because this branch is no longer under development.
